### ISSUE TITLE
Resolve a Promise Asynchronously

### ISSUE THREAD

**Issue by vpaladino778** [nodejs/node-addon-api#607]
I'm attempting to resolve a promise in an asynchronous fashion via callback function. The following code will communicate what I'm trying to do, but it does not run.

```c++
Napi::Promise::Deferred *myPromise;
Napi::Env* myPromiseEnv;

Napi::Promise myJSPromiseFunc(const Napi::CallbackInfo& info) {
    Napi::Promise::Deferred prom = Napi::Promise::New(info.Env());

    myPromise = &prom;
    myPromiseEnv = &info.Env();
    // After this is called, myCallbackFunction() will eventually get called after doing some processing
    triggerCallbackFunctionLater();
    return prom.Promise();
}

// This will get called later, it's intended to resolve the promise here
void myCallbackFunction(String result) {
    Napi::Object resolveObject = Napi::Object::New(*myPromiseEnv);
    resolveObject.Set(Napi::String::New(*myPromiseEnv, "result"), Napi::String::New(*myPromiseEnv, result));
    myPromise->Resolve(resolveObject);
}
```

So after testing something like this out, i realized that when `myCallbackFunction()` runs, it will crash because I found out that the Env variable cannot be passed around all willy nilly like that. A

After reading through some of the documentation, i found [this](https://github.com/nodejs/node-addon-api/blob/master/doc/threadsafe_function.md) and I was able to concoct a convoluted solution using a global ThreadSafeFunction object along with a global promise object. I passed in a blank Napi::Function into the ThreadSafeFunction (Because i don't need a javascript callback), and then made a new CallbackFunction which does a nonblocking call on my threadsafe callback. and then on the function i pass to the non blocking call, i resolve the promise. It's very convoluted and doesn't seem to be the right way to do it.

So In summary, i'd like to have something like this:
```c++
Napi::Promise myJSPromiseFunc(const Napi::CallbackInfo& info) {
    Napi::Promise::Deferred prom = Napi::Promise::New(info.Env());
    // myAsynchFunc() will get run asynchronously and then the promise will eventually be resolved within myAsyncFunc
    Napi::RunAsynchronously(myAsyncFunc, prom, info.Env());
    return prom.Promise();
}
```

I was wondering if anyone else had an idea of a better way to achieve something like this. Any input or feedback would be greatly appreciated. Thank you!

**Comment by vpaladino778** [comment]
I've done some more reading into the documentation and came across the [AsyncWorker](https://github.com/nodejs/node-addon-api/blob/master/doc/async_worker.md) class. This appears to be along the lines of what I'm looking for. Now i'm just having trouble resolving the promise inside the class. Again having issues with Env.

**Comment by KevinEady** [comment]
Hi @vpaladino778 ,

If you're using `AsyncWorker`, you can override the `OnOK()` method (which executes on the main thread), and access `AsyncWorker::Env()` there.

You can't access `Env` in your [`Execute`](https://github.com/nodejs/node-addon-api/blob/master/doc/async_worker.md#execute) method like you mentioned in your edit, see docs:

> As the method is not running on the main event loop, it must avoid calling any methods from node-addon-api or running any code that might invoke JavaScript.

**Comment by KevinEady** [comment]
@vpaladino778 maybe this example helps?

https://github.com/nodejs/node-addon-examples/issues/85#issuecomment-551142933

**Comment by vpaladino778** [comment]
@KevinEady Thank you so much for your response. You helped me solve exactly what i was looking for. Also that example is exactly what I needed. Thank you so much for your help, it was very much appreciated!

**Comment by SurienDG** [comment]
@vpaladino778 If you are working with promises a lot with your function this may be a helpful module to look at: https://github.com/SurienDG/NAPI-Thread-Safe-Promise/
